server/queue/queue-rabbitmq/src/main/java/org/apache/james/queue/rabbitmq/Dequeuer.java

```diff
@@ -0,0 +1,132 @@
+ * Licensed to the Apache Software Foundation (ASF) under one   *
+ * or more contributor license agreements.  See the NOTICE file *
+ * distributed with this work for additional information        *
+ * regarding copyright ownership.  The ASF licenses this file   *
+ * to you under the Apache License, Version 2.0 (the            *
+ * "License"); you may not use this file except in compliance   *
+ * with the License.  You may obtain a copy of the License at   *
+ *                                                              *
+ *   http://www.apache.org/licenses/LICENSE-2.0                 *
+ *                                                              *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.queue.rabbitmq;
+
+import static org.apache.james.queue.api.MailQueue.DEQUEUED_METRIC_NAME_PREFIX;
+
+import java.io.IOException;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executors;
+import java.util.function.Consumer;
+import java.util.function.Function;
+
+import org.apache.james.metrics.api.Metric;
+import org.apache.james.metrics.api.MetricFactory;
+import org.apache.james.queue.api.MailQueue;
+import org.apache.mailet.Mail;
+
+import com.github.fge.lambdas.Throwing;
+import com.github.fge.lambdas.consumers.ThrowingConsumer;
+import com.nurkiewicz.asyncretry.AsyncRetryExecutor;
+import com.rabbitmq.client.GetResponse;
+
+class Dequeuer {
+    private static class NoMailYetException extends RuntimeException {
+    }
+
+    private static class RabbitMQMailQueueItem implements MailQueue.MailQueueItem {
+        private final Consumer<Boolean> ack;
+        private final Mail mail;
+
+        private RabbitMQMailQueueItem(Consumer<Boolean> ack, Mail mail) {
+            this.ack = ack;
+            this.mail = mail;
+        }
+
+        @Override
+        public Mail getMail() {
+            return mail;
+        }
+
+        @Override
+        public void done(boolean success) throws MailQueue.MailQueueException {
+            ack.accept(success);
+        }
+    }
+
+    private static final int TEN_MS = 10;
+
+    private final MailQueueName name;
+    private final RabbitClient rabbitClient;
+    private final Function<MailReferenceDTO, Mail> mailLoader;
+    private final Metric dequeueMetric;
+    private final MailReferenceSerializer mailReferenceSerializer;
+
+    Dequeuer(MailQueueName name, RabbitClient rabbitClient, Function<MailReferenceDTO, Mail> mailLoader, MailReferenceSerializer serializer, MetricFactory metricFactory) {
```

Comment:
Aren't we enforcing `private`/`public` keywords?

Reply:
Well, because the Dequeuer is just logic extracted from RabbitMQ mail queue for readability/separation of concerns.

However, that class makes no sense alone: it thus is not exposed, not standalone tested, etc... Hence visibility reduction

CF recent chat discussion on the topic.

Reply:
Well, because the Dequeuer is just logic extracted from RabbitMQ mail queue for readability/separation of concerns.

However, that class makes no sense alone: it thus is not exposed, not standalone tested, etc... Hence visibility reduction

CF recent chat discussion on the topic.

Reply:
Well, because the Dequeuer is just logic extracted from RabbitMQ mail queue for readability/separation of concerns.

However, that class makes no sense alone: it thus is not exposed, not standalone tested, etc... Hence visibility reduction

CF recent chat discussion on the topic.